### DISCUSSION TITLE
Apply mlmg to compute fluxes and gradient

### DISCUSSION THREAD

**Discussion by ZiyangHuang** [AMReX-Codes/amrex#3492] [Q&A] [answered]
Greetings,

If my understanding is correct, I need to first call ```mlmg.solve(...)``` to obtain the solution of a linear solve, and then call ```mlmg.getFluxes(...)``` and ```mlmg.getGradSolution(...)``` to obtain the fluxes and gradient. 

I am aware that there is another member function:
```
void amrex::MLMGT< MF >::getFluxes (  const Vector< Array< AMF *, AMREX_SPACEDIM > > &  a_flux,
                                                                 const Vector< AMF * > &  a_sol,
                                                                Location  a_loc = Location::FaceCenter
                                                              )
```
to compute the fluxes with input ```a_sol```. Can I use this function to compute fluxes related to a ```MLLinOp``` without the need of calling ```mlmg.solve(...)``` beforehand?

However, I could not find a similar version of member function for ```getGradSolution```. Does it mean that I always need to call ```mlmg.solve()``` before I compute the gradient?

Thanks!

**Comment by WeiqunZhang** [comment, marked as answer]
Yes, you have to call `solve`. There is no guarantee that `getFluxes` and `getGradSolution` could work without calling `solve` first. However, you could call `solve` with a huge tolerance that no iteration will be needed. You could also call `LPInfo::setMaxCoarseningLevel(0)` to reduce the setup cost of the `LinOp`.
